### PR TITLE
Add Emscripten system and emcc toolset support.

### PR DESCRIPTION
**Pull request by tritao**
**What does this PR do?**

Thanks for the contribution! Please provide a concise description of the problem this request solves.

**How does this PR change Premake's behavior?**

Are there any breaking changes? Will any existing behavior change?

**Anything else we should know?**

Add any other context about your changes here.

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [x] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [x] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

*You can now [support Premake on our OpenCollective](https://opencollective.com/premake). Your contributions help us spend more time responding to requests like these!*

### CODE REVIEW

File: src/tools/emcc.lua [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+-- @param cfg
+--    The project configuration.
+-- @param dirs
+--    An array of include file search directories; as an array of
+--    string values.
+-- @return
+--    An array of symbols with the appropriate flag decorations.
+--
+
+	function emcc.getincludedirs(cfg, dirs)
```
Comment (Jarod42) [inline src/tools/emcc.lua:136]:
```suggestion
	function emcc.getincludedirs(cfg, includedirs, externalincludedirs, frameworkdirs, includedirsafter)
```

File: src/tools/emcc.lua [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+--    default value should be used.
+--
+
+	emcc.tools = {
+		cc = "emcc",
+		cxx = "em++",
+		ar = "emar"
+	}
+
+	function emcc.gettoolname(cfg, tool)
```
Comment (Jarod42) [inline src/tools/emcc.lua:279]: Is there version to handle too for emcc (as clang-14, gcc-9)?

File: src/base/os.lua [hunk 3]
```diff
@@ -827,16 +827,17 @@
...
-		["ios"]      = { "ios",     "darwin", "posix", "mobile" },
-		["linux"]    = { "linux",   "posix", "desktop" },
-		["macosx"]   = { "macosx",  "darwin", "posix", "desktop" },
-		["solaris"]  = { "solaris", "posix", "desktop" },
-		["uwp"]      = { "uwp", "windows", "desktop" },
-		["windows"]  = { "windows", "win32", "desktop" },
+		["aix"]        = { "aix",     "posix", "desktop" },
+		["android"]    = { "android", "mobile" },
+		["bsd"]        = { "bsd",     "posix", "desktop" },
+		["emscripten"] = { "emscripten", "web" },
```
Comment (redorav) [inline src/base/os.lua:833]: Not sure why it's marking this as many changed lines?

File: src/base/oven.lua [hunk 4]
```diff
@@ -538,6 +538,9 @@
 		local system = os.target()
 		local architecture = os.targetarch()
 		local toolset = p.action.current().toolset
+		if type(toolset) == "function" then
+			toolset = toolset()
```
Comment (redorav) [inline src/base/oven.lua:542]: Why does this need to be a function and not evaluated in place as it used to be?

File: contrib/libzip/premake5.lua [hunk 5]
```diff
@@ -11,6 +11,10 @@ project "zip-lib"
 		"**.c"
 	}
 
+	filter "system:linux"
+		defines { "HAVE_SSIZE_T_LIBZIP", "HAVE_CONFIG_H" }
```
Comment (redorav) [inline contrib/libzip/premake5.lua:15]: Is this relevant to this commit too?

File: website/docs/system.md [hunk 6]
```diff
@@ -23,6 +24,8 @@ If no system is specified, Premake will identify and target the current operatin
 * windows
 * xbox360
 
+To note: `emscripten` at the moment is only supported for the `gmake` and `gmake2` actions.
```
Comment (nickclark2016) [inline website/docs/system.md:27]: For uniformity with the rest of the docs, use a separate "Availability" header and make any notes beneath that. See https://raw.githubusercontent.com/premake/premake-core/refs/heads/master/website/docs/debugformat.md

Comment (Jarod42) [reply website/docs/system.md:27]: Pretty sure that others values are not cross-generator neither.

**Review by Jarod42 — APPROVED** [review]

File: website/docs/system.md (hunk 6 above)
Comment (nickclark2016) [reply website/docs/system.md:27]: We should probably clean them up at some point, then.
